File: app/src/main/java/com/asteph11/finalproject/models/data/Player.java
package com.asteph11.finalproject.models.data;

public class Player {

    private String name;
    private Grid grid;

    private int numShips = 0;
    private int hits, misses;

    public Player(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public boolean setShip(int x, int y) {
        if(grid == null) {
            grid = new Grid();
        }

        if(grid.getStatusAt(x, y) != Grid.Status.SHIP) {
            if(!isGridReady()) {
                grid.setStatus(x, y, Grid.Status.SHIP);
                numShips++;
                return true;
            } else {
                return false;
            }
        } else {
            grid.setStatus(x, y, Grid.Status.EMPTY);
            numShips--;
            if(numShips < 0) {
                numShips = 0;
            }
            return true;
        }
    }

    /**
     * Sets the state of the player's own grid upon receiving fire
     * @param x horizontal coord
     * @param y vertical coord
     * @return -1 already hit/retry, 0 empty no hit, 1 ship hit
     */
    public int receiveFire(int x, int y) throws Exception {
        if(grid == null) {
            throw new Exception("Player grid is null!");
        }

        int value = 0;

        Grid.Status status = grid.getStatusAt(x, y);
        switch(status) {
            case HIT: case MISS: {
                value = -1;
                break;
            }
            case SHIP: {
                grid.setStatus(x, y, Grid.Status.HIT);
                value = 1;
                break;
            }
            case EMPTY: {
                grid.setStatus(x, y, Grid.Status.MISS);
                value = 0;
                break;
            }
        }

        return value;
    }

    public boolean isReady() {
        return grid != null;
    }

    public Grid getGrid() {
        return grid;
    }

    public boolean isDefeated() {
        boolean isDefeated = true;
        for(int y = 0; y < Grid.DIMENSIONS; y++) {
            for(int x = 0; x < Grid.DIMENSIONS; x++) {
                if(grid.getStatusAt(x, y, Grid.Status.SHIP)) {
                    isDefeated = false;
                }
            }
        }
        return isDefeated;
    }

    public void setName(String nameInput) {
        this.name = nameInput;
    }

    public boolean isGridReady() {
        return numShips >= Grid.MAX_NUM_SHIPS;
    }

    public void addHit() {
        hits++;
    }

    public void addMiss() {
        misses++;
    }

    public int getHits() {
        return hits;
    }

    public int getMisses() {
        return misses;
    }
}

File: app/src/main/java/com/asteph11/finalproject/models/data/Player.kt
package com.asteph11.finalproject.models.data

import java.lang.Exception
import kotlin.Throws

class Player(var name: String) {
    var grid: Grid? = null
        private set
    private var numShips = 0
    var hits = 0
        private set
    var misses = 0
        private set

    fun setShip(x: Int, y: Int): Boolean {
        if (grid == null) {
            grid = Grid()
        }
        return if (grid!!.getStatusAt(x, y) != Grid.Status.SHIP) {
            if (!isGridReady) {
                grid!!.setStatus(x, y, Grid.Status.SHIP)
                numShips++
                true
            } else {
                false
            }
        } else {
            grid!!.setStatus(x, y, Grid.Status.EMPTY)
            numShips--
            if (numShips < 0) {
                numShips = 0
            }
            true
        }
    }

    /**
     * Sets the state of the player's own grid upon receiving fire
     * @param x horizontal coord
     * @param y vertical coord
     * @return -1 already hit/retry, 0 empty no hit, 1 ship hit
     */
    @Throws(Exception::class)
    fun receiveFire(x: Int, y: Int): Int {
        if (grid == null) {
            throw Exception("Player grid is null!")
        }
        var value = 0
        val status = grid!!.getStatusAt(x, y)
        value = when (status) {
            Grid.Status.HIT, Grid.Status.MISS -> {
                -1
            }
            Grid.Status.SHIP -> {
                grid!!.setStatus(x, y, Grid.Status.HIT)
                1
            }
            Grid.Status.EMPTY -> {
                grid!!.setStatus(x, y, Grid.Status.MISS)
                0
            }
        }
        return value
    }

    val isReady: Boolean
        get() = grid != null
    val isDefeated: Boolean
        get() {
            var isDefeated = true
            for (y in 0 until Grid.DIMENSIONS) {
                for (x in 0 until Grid.DIMENSIONS) {
                    if (grid!!.getStatusAt(x, y, Grid.Status.SHIP)) {
                        isDefeated = false
                    }
                }
            }
            return isDefeated
        }
    val isGridReady: Boolean
        get() = numShips >= Grid.MAX_NUM_SHIPS

    fun addHit() {
        hits++
    }

    fun addMiss() {
        misses++
    }
}